### PR TITLE
gh-94597: deprecate asyncio.set_event_loop_policy

### PR DESCRIPTION
**Pull request by graingert**
----
:books: Documentation preview :books:: https://cpython-previews--110728.org.readthedocs.build/

* Issue: gh-94597

### CODE REVIEW

**Review by cjw296 — COMMENTED** [review]
No objection to the unittest or AsyncMock changes!

File: Doc/library/asyncio-policy.rst
Comment (gvanrossum) [inline Doc/library/asyncio-policy.rst]: We should review the deprecation once we've reached that point:
```suggestion
be removed in Python 3.15 or later.
```

File: Lib/asyncio/events.py
Comment (gvanrossum) [inline Lib/asyncio/events.py]: Either "may be removed" or "3.15 or later"?

File: Lib/test/test_asyncio/test_events.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_events.py]: `cm` unused?
```suggestion
        with self.assertWarns(DeprecationWarning):
            asyncio.set_event_loop_policy(policy)
```

File: Misc/NEWS.d/next/Library/2023-10-11-15-55-53.gh-issue-94597.qXo8Q4.rst
Comment (gvanrossum) [inline Misc/NEWS.d/next/Library/2023-10-11-15-55-53.gh-issue-94597.qXo8Q4.rst]: Misc/NEWS entries should be capitalized and punctuated properly (they are formatted as paragraphs ReST text somewhere).
```suggestion
Deprecate :func:`asyncio.set_event_loop_policy`.
```

**Review by gvanrossum — COMMENTED** [review]
Just a few nits, basically LG. That `tearDownModule()` pattern is annoying! :-)

I assume that deprecating `get_event_loop_policy()` will follow? Are there other APIs? I'm guessing we also want to avoid instantiating or subclassing the standard event loop policy classes (I expect this will turn up some significant uses of them).

Before we merge this we may need to bring this up on discuss.python.org (in the asyncio category) to see if there are significant objections to the proposed deprecation schedule.

**Comment by graingert** [comment]
> Just a few nits, basically LG. That `tearDownModule()` pattern is annoying! :-)

I think we can start getting rid of tearDownModules using #110774 

> I assume that deprecating `get_event_loop_policy()` will follow? Are there other APIs? I'm guessing we also want to avoid instantiating or subclassing the standard event loop policy classes (I expect this will turn up some significant uses of them).

Yep currently that's a bit more complicated because get/set_event_loop calls get_event_loop_policy and afaik we don't want to make get_event_loop an alias of get_running_loop anymore

File: Lib/test/libregrtest/save_env.py
Comment (willingc) [inline Lib/test/libregrtest/save_env.py:99]: Is line 99 still necessary?

File: Doc/library/asyncio-policy.rst
Comment (willingc) [reply Doc/library/asyncio-policy.rst]: Do we want to give a reason as to why the policy system is being deprecated?

**Review by willingc — COMMENTED** [review]
Thanks for the PR. I've left a few questions for you.

File: Doc/library/asyncio-policy.rst
Comment (gvanrossum) [reply Doc/library/asyncio-policy.rst]: Good question. I don't think there's anyone among the (current or past) asyncio maintainers thinks the policy system was a good idea -- it's too complex with not enough benefits. But do we need to explain this in the docs? I don't believe we usually document the reason for a deprecation.

**Comment by hugovk** [comment]
Please also mention the deprecation in What's New in Python 3.13.

Because it's to be removed in 3.15 or later it could be listed under either [Pending Removal in Python 3.15](https://docs.python.org/3.13/whatsnew/3.13.html#pending-removal-in-python-3-15) or [Pending Removal in Future Versions](https://docs.python.org/3.13/whatsnew/3.13.html#pending-removal-in-future-versions).

I suggest the former, as it's more likely to be noticed and action taken, and include the "3.15 or later" wording. We can update later if needed.

File: Lib/test/libregrtest/save_env.py
Comment (graingert) [inline Lib/test/libregrtest/save_env.py]:
```suggestion
```

File: Lib/test/libregrtest/save_env.py
Comment (graingert) [reply Lib/test/libregrtest/save_env.py:99]: no, I've removed it

File: Lib/asyncio/events.py
Comment (graingert) [reply Lib/asyncio/events.py]:
```suggestion
                         "{name!r} is deprecated as of Python 3.13 and will be "
                         "removed in Python {remove} or later.", remove=(3, 15))
```

File: Doc/whatsnew/3.13.rst
Comment (hugovk) [inline Doc/whatsnew/3.13.rst]: 3.15 or later?

```suggestion
  rest of the asyncio policy system in Python 3.15 or later.
```

**Comment by graingert** [comment]
> Before we merge this we may need to bring this up on discuss.python.org (in the asyncio category) to see if there are significant objections to the proposed deprecation schedule.

discuss post posted here: https://discuss.python.org/t/removing-the-asyncio-policy-system-asyncio-set-event-loop-policy-in-python-3-15/37553

**Review by gvanrossum — APPROVED** [review]
Okay, I'm happy with this. Let's wait a bit if anyone brings up substantial objections in the [Discourse thread](https://discuss.python.org/t/removing-the-asyncio-policy-system-asyncio-set-event-loop-policy-in-python-3-15/37553).

**Review by willingc — APPROVED** [review]
Bravo @graingert for simplifying. ✨

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]
Please also add a notice under `asyncio` changes, most users like myself won't look under 3.15 removals for this change. Thanks

**Comment by graingert** [comment]
> Please also add a notice under `asyncio` changes, most users like myself won't look under 3.15 removals for this change. Thanks

Can you link me to the location in the docs that you mean please?

**Comment by graingert** [comment]
> Please also add a notice under `asyncio` changes, most users like myself won't look under 3.15 removals for this change. Thanks

Can you link me to the location in the docs that you mean please?

**Comment by kumaraditya303** [comment]
> Can you link me to the location in the docs that you mean please?

Add `asyncio` entry in https://docs.python.org/3.13/whatsnew/3.13.html#deprecated for `asyncio`

**Comment by hugovk** [comment]
That sounds fine, then let's move it from the "Pending Removal in Python 3.15" section to this one.

**Comment by graingert** [comment]
I have made the requested changes; please review again

**Review by gvanrossum — APPROVED** [review]
LGTM, but let's wait until the dust in the Discourse thread settles.

**Comment by 1st1** [comment]
Happy to review and hopefully merge this during the core sprint.

**Comment by graingert** [comment]
@kumaraditya303 is doing this now

**Pull request closed without merging**
